### PR TITLE
added executable and environment parameters for jibDockerBuild

### PR DESCRIPTION
**Pull request by gintautassulskus**
Addresses https://github.com/GoogleContainerTools/jib/issues/468

jibDockerBuild can now be configured with two optional properties:
```
jibDockerBuild {
    dockerExecutable = "/optional/path/to/docker/"

    def envVars = [:]
        envVars["DOCKER_TLS_VERIFY"] = "1"
        envVars["DOCKER_HOST"] = "tcp://IP:PORT"
        envVars["DOCKER_CERT_PATH"] = "/path/to/.minikube/certs"
        envVars["DOCKER_API_VERSION"] = "1.35"
    dockerEnvironment = envVars
}
```

Also, reformatted source code to follow the suggested [Oracle conventions](https://www.oracle.com/technetwork/java/javase/documentation/codeconventions-136091.html) (4-space indentations). Hence so many changed lines. Now not sure any more whether this was a good idea.

### CODE REVIEW

**Comment by briandealwis** [comment]
> Also, reformatted source code to follow the suggested Oracle conventions (4-space indentations). Hence so many changed lines. Now not sure any more whether this was a good idea.

It's not documented in the CONTRIBUTING.md, but we use the [Google Java Style Guide](https://google.github.io/styleguide/javaguide.html).

**Comment by gintautassulskus** [comment]
@briandealwis, my bad. Addressed. 
The update should come up shortly.

**Comment by briandealwis** [comment]
Thanks @gintautassulskus — if you run `./build.sh`, it will reformat and run basic tests.

File: jib-core/src/main/java/com/google/cloud/tools/jib/docker/DockerClient.java [hunk 1]
```diff
@@ -20,28 +20,31 @@
...
 import java.util.function.Function;
 
-/** Calls out to the {@code docker} CLI. */
+/**
+ * Calls out to the {@code docker} CLI.
+ */
 public class DockerClient {
 
   private static final String DEFAULT_DOCKER_CLIENT = "docker";
+  private static final Map<String, String> DEFAULT_DOCKER_ENVIRONMENT = Collections.emptyMap();
```
Comment (briandealwis) [inline jib-core/src/main/java/com/google/cloud/tools/jib/docker/DockerClient.java]: So this results in a change in behaviour as previously the docker command would have inherited the current environment.  This may actually break existing users who are using `minikube docker-env`.  I this should should default to `null` and the `defaultProcessBuilderFactory` should only set the environment if non-null.

File: jib-core/src/main/java/com/google/cloud/tools/jib/api/DockerDaemonImage.java [hunk 2]
```diff
@@ -56,8 +62,11 @@ public static DockerDaemonImage named(String imageReference)
 
   private final ImageReference imageReference;
   private Path dockerExecutable = Paths.get("docker");
+  private Map<String, String> dockerEnvironment = Collections.emptyMap();
```
Comment (briandealwis) [inline jib-core/src/main/java/com/google/cloud/tools/jib/api/DockerDaemonImage.java]: As I mention in the `DockerClient` change, this is a change in behaviour as it will cause the environment to be replaced.

**Comment by briandealwis** [comment]
It would be helpful to add a test to `DockerClientTest` too.

File: jib-core/src/main/java/com/google/cloud/tools/jib/docker/DockerClient.java (hunk 1 above)
Comment (gintautassulskus) [reply jib-core/src/main/java/com/google/cloud/tools/jib/docker/DockerClient.java]: The default `DEFAULT_DOCKER_ENVIRONMENT` value was not there before. If no environment variable is passed, this is then used as a default one. Then, I simply add `dockerEnvironment` to the map of inherited ENV variables:
```
      ProcessBuilder pb = new ProcessBuilder(dockerCommand);

      Map<String, String> env = pb.environment();
      env.putAll(dockerEnvironment);
```

Therefore, I do not think this will introduce any breaking changes. 

You could argue, however, whether there is a point of passing an empty list as a default value. In this context, I kept consistent with the `dockerExecutable` configuration.

Ideally, I think, `DockerDaemonImage` should have no configuration defaults and call `newClient()` or `newClient(null, null)`. `DockerClient` would then check the parameters and would set defaults in case of `null`'s.

File: jib-core/src/main/java/com/google/cloud/tools/jib/api/DockerDaemonImage.java (hunk 2 above)
Comment (gintautassulskus) [reply jib-core/src/main/java/com/google/cloud/tools/jib/api/DockerDaemonImage.java]: Same as above. In my opinion, `dockerExecutable` and `dockerEnvironment` should be removed from this class and left in `DockerClient` in order to eliminate duplication.

File: jib-core/src/main/java/com/google/cloud/tools/jib/docker/DockerClient.java (hunk 1 above)
Comment (briandealwis) [reply jib-core/src/main/java/com/google/cloud/tools/jib/docker/DockerClient.java]: > Therefore, I do not think this will introduce any breaking changes.

You're right — I misread. My apologies.

**Comment by briandealwis** [comment]
@gintautassulskus it would be helpful for the team to better understand your context.  Would you mind describing how this change and your minikube PR fit into your processes — perhaps on #468?  Have you looked at [Skaffold](https://github.com/GoogleContainerTools/skaffold)?

**Comment by gintautassulskus** [comment]
> It would be helpful to add a test to `DockerClientTest` too.

Please check the latest commit. The most straightforward approach to test DockerClient that I have come up requires exposing the `processBuilderFactory`:
```
  @VisibleForTesting
  Function<List<String>, ProcessBuilder> getProcessBuilderFactory() {
    return processBuilderFactory;
  }
```
Let me know if this is acceptable. I have noticed a similar approach throughout the project.

File: jib-core/src/main/java/com/google/cloud/tools/jib/docker/DockerClient.java (hunk 1 above)
Comment (gintautassulskus) [reply jib-core/src/main/java/com/google/cloud/tools/jib/docker/DockerClient.java]: No prob. Thanks for the feedback!

**Comment by gintautassulskus** [comment]
> @gintautassulskus it would be helpful for the team to better understand your context. Would you mind describing how this change and your minikube PR fit into your processes — perhaps on #468? Have you looked at [Skaffold](https://github.com/GoogleContainerTools/skaffold)?

Thanks, @briandealwis, Skaffold is something new to me. Sounds very promising.

**Comment by chanseokoh** [comment]
@gintautassulskus thanks for putting up this. Our team had an internal discussion on this topic of how best this feature can be supported. We will finalize the design after doing some more discussions and research, but it may take some time as our team members are busy with various events. Please be patient, and we will get back to you later.

**Comment by chanseokoh** [comment]
We had an internal discussion, and basically, the configurable-environment-variable approach makes sense. We just decided to use a slightly different config names:

```
jibDockerBuild.dockerClient.executable = ...
jibDockerBuild.dockerClient.environment = [key, 'value', ...]
```

It'd be nice if you can update the `CHANGLOG.md` in this PR too, but we can do it too if that slows you down.

File: jib-core/src/main/java/com/google/cloud/tools/jib/docker/DockerClient.java [hunk 3]
```diff
@@ -25,23 +25,23 @@
 import java.io.OutputStream;
 import java.nio.charset.StandardCharsets;
 import java.nio.file.Path;
-import java.util.ArrayList;
-import java.util.Arrays;
-import java.util.List;
+import java.util.*;
```
Comment (chanseokoh) [inline jib-core/src/main/java/com/google/cloud/tools/jib/docker/DockerClient.java]: Should be expanded per our style guide.

File: jib-core/src/main/java/com/google/cloud/tools/jib/docker/DockerClient.java [hunk 4]
```diff
@@ -62,12 +76,18 @@ public static DockerClient newClient(Path dockerExecutable) {
...
   private static Function<List<String>, ProcessBuilder> defaultProcessBuilderFactory(
-      String dockerExecutable) {
+      String dockerExecutable, Map<String, String> dockerEnvironment) {
     return dockerSubCommand -> {
       List<String> dockerCommand = new ArrayList<>(1 + dockerSubCommand.size());
       dockerCommand.add(dockerExecutable);
       dockerCommand.addAll(dockerSubCommand);
-      return new ProcessBuilder(dockerCommand);
+
+      ProcessBuilder pb = new ProcessBuilder(dockerCommand);
```
Comment (chanseokoh) [inline jib-core/src/main/java/com/google/cloud/tools/jib/docker/DockerClient.java]: nit: let's say `ProcessBuilder processBuilder = new ...`.

File: jib-core/src/main/java/com/google/cloud/tools/jib/docker/DockerClient.java [hunk 5]
```diff
@@ -62,12 +76,18 @@ public static DockerClient newClient(Path dockerExecutable) {
...
+
+      Map<String, String> env = pb.environment();
```
Comment (chanseokoh) [inline jib-core/src/main/java/com/google/cloud/tools/jib/docker/DockerClient.java]: nit: `... environment = ...`

File: jib-core/src/test/java/com/google/cloud/tools/jib/docker/DockerClientTest.java [hunk 6]
```diff
@@ -16,6 +16,8 @@
 
 package com.google.cloud.tools.jib.docker;
 
+import static org.mockito.Mockito.when;
```
Comment (chanseokoh) [inline jib-core/src/test/java/com/google/cloud/tools/jib/docker/DockerClientTest.java]: nit: in our repo, we just do `Mockito.when(...)` without static import.

File: jib-core/src/test/java/com/google/cloud/tools/jib/docker/DockerClientTest.java [hunk 7]
```diff
@@ -176,11 +181,37 @@ public void testTag() throws InterruptedException, IOException, InvalidImageRefe
...
             });
-    Mockito.when(mockProcess.waitFor()).thenReturn(0);
+    when(mockProcess.waitFor()).thenReturn(0);
 
     testDockerClient.tag(ImageReference.of(null, "original", null), ImageReference.parse("new"));
   }
 
+  @Test
+  public void testDefaultClient() {
+    DockerClient dc = DockerClient.newClient();
```
Comment (chanseokoh) [inline jib-core/src/test/java/com/google/cloud/tools/jib/docker/DockerClientTest.java]: nit: `DockerClient client = ...`

File: jib-core/src/test/java/com/google/cloud/tools/jib/docker/DockerClientTest.java [hunk 8]
```diff
@@ -176,11 +181,37 @@ public void testTag() throws InterruptedException, IOException, InvalidImageRefe
...
+    Assert.assertEquals(
+        dc.getProcessBuilderFactory().apply(Collections.emptyList()).command(),
+        Collections.singletonList("docker"));
```
Comment (chanseokoh) [inline jib-core/src/test/java/com/google/cloud/tools/jib/docker/DockerClientTest.java]: In our repo, we reverse the order so that `assertEquals(expected, actual)`.

File: jib-gradle-plugin/README.md [hunk 9]
```diff
@@ -108,6 +108,21 @@ eval $(minikube docker-env)
 gradle jibDockerBuild
 ```
 
+Alternatively, the environment can be configured in the task with two optional parameters:
```
Comment (chanseokoh) [inline jib-gradle-plugin/README.md]: We don't update `README` until we actually make a release, because it may confuse the users of the current latest version. Should be removed here. It'd be nice to put up a separate PR.

File: jib-gradle-plugin/src/main/java/com/google/cloud/tools/jib/gradle/BuildDockerTask.java [hunk 10]
```diff
@@ -135,4 +148,24 @@ public BuildDockerTask setJibExtension(JibExtension jibExtension) {
     this.jibExtension = jibExtension;
     return this;
   }
+
+  @Nullable
+  public Map<String, String> getDockerEnvironment() {
```
Comment (chanseokoh) [inline jib-gradle-plugin/src/main/java/com/google/cloud/tools/jib/gradle/BuildDockerTask.java]: Is this method necessary?

File: jib-gradle-plugin/src/main/java/com/google/cloud/tools/jib/gradle/BuildDockerTask.java [hunk 11]
```diff
@@ -135,4 +148,24 @@ public BuildDockerTask setJibExtension(JibExtension jibExtension) {
...
+    return dockerEnvironment;
+  }
+
+  @Input
+  public void setDockerEnvironment(@Nullable Map<String, String> dockerEnvironment) {
+    this.dockerEnvironment = dockerEnvironment;
+  }
+
+  @Nullable
+  public Path getDockerExecutable() {
```
Comment (chanseokoh) [inline jib-gradle-plugin/src/main/java/com/google/cloud/tools/jib/gradle/BuildDockerTask.java]: Is this method necessary?

File: jib-gradle-plugin/src/main/java/com/google/cloud/tools/jib/gradle/BuildDockerTask.java [hunk 12]
```diff
@@ -135,4 +148,24 @@ public BuildDockerTask setJibExtension(JibExtension jibExtension) {
...
     return this;
   }
+
+  @Nullable
+  public Map<String, String> getDockerEnvironment() {
+    return dockerEnvironment;
+  }
+
+  @Input
+  public void setDockerEnvironment(@Nullable Map<String, String> dockerEnvironment) {
```
Comment (chanseokoh) [inline jib-gradle-plugin/src/main/java/com/google/cloud/tools/jib/gradle/BuildDockerTask.java]: Just a reminder: as I said, we will change the config name/structure.

File: jib-core/src/main/java/com/google/cloud/tools/jib/api/DockerDaemonImage.java (hunk 2 above)
Comment (chanseokoh) [reply jib-core/src/main/java/com/google/cloud/tools/jib/api/DockerDaemonImage.java]: Yeah, I think it makes sense to remove both `setDockerExecutable` and `setDockerEnvironment` and instead have `setDockerClient()`.

File: jib-core/src/main/java/com/google/cloud/tools/jib/docker/DockerClient.java [hunk 13]
```diff
@@ -79,6 +99,11 @@ public static DockerClient newClient(Path dockerExecutable) {
     this.processBuilderFactory = processBuilderFactory;
   }
 
+  @VisibleForTesting
+  Function<List<String>, ProcessBuilder> getProcessBuilderFactory() {
```
Comment (chanseokoh) [inline jib-core/src/main/java/com/google/cloud/tools/jib/docker/DockerClient.java]: nit: non-public methods go below public in code.

BTW, instead of having this new getter, is it possible to make `defaultProcessBuilderFactory()` `@VisibleForTesting`?

File: jib-core/src/test/java/com/google/cloud/tools/jib/docker/DockerClientTest.java [hunk 14]
```diff
@@ -176,11 +181,37 @@ public void testTag() throws InterruptedException, IOException, InvalidImageRefe
...
+  @Test
+  public void testCustomClient() {
+    Path path = Paths.get("/path.to/docker");
+    Map<String, String> environment = new HashMap<String, String>();
+    environment.putAll(System.getenv());
+    environment.putAll(Collections.singletonMap("Key1", "Value1"));
+
+    DockerClient dc = DockerClient.newClient(path, environment);
+    Assert.assertEquals(
+        dc.getProcessBuilderFactory().apply(Collections.emptyList()).command(),
```
Comment (chanseokoh) [inline jib-core/src/test/java/com/google/cloud/tools/jib/docker/DockerClientTest.java]: nit: the `dc` name, and the order of assert arguments.

File: jib-core/src/test/java/com/google/cloud/tools/jib/docker/DockerClientTest.java [hunk 15]
```diff
@@ -176,11 +181,37 @@ public void testTag() throws InterruptedException, IOException, InvalidImageRefe
...
+        dc.getProcessBuilderFactory().apply(Collections.emptyList()).environment(),
+        System.getenv());
+  }
+
+  @Test
+  public void testCustomClient() {
+    Path path = Paths.get("/path.to/docker");
+    Map<String, String> environment = new HashMap<String, String>();
+    environment.putAll(System.getenv());
+    environment.putAll(Collections.singletonMap("Key1", "Value1"));
```
Comment (chanseokoh) [inline jib-core/src/test/java/com/google/cloud/tools/jib/docker/DockerClientTest.java]: Why not just `put("Key1", "Value1")`?

File: jib-core/src/test/java/com/google/cloud/tools/jib/docker/DockerClientTest.java [hunk 16]
```diff
@@ -176,11 +181,37 @@ public void testTag() throws InterruptedException, IOException, InvalidImageRefe
...
+        Collections.singletonList("docker"));
+    Assert.assertEquals(
+        dc.getProcessBuilderFactory().apply(Collections.emptyList()).environment(),
+        System.getenv());
+  }
+
+  @Test
+  public void testCustomClient() {
+    Path path = Paths.get("/path.to/docker");
+    Map<String, String> environment = new HashMap<String, String>();
```
Comment (chanseokoh) [inline jib-core/src/test/java/com/google/cloud/tools/jib/docker/DockerClientTest.java]: `... = HashMap<>(System.getenv())`?

At least remove `HashMap<String, String>()` --> `HashMap<>()`.

File: jib-core/src/main/java/com/google/cloud/tools/jib/docker/DockerClient.java (hunk 3 above)
Comment (gintautassulskus) [reply jib-core/src/main/java/com/google/cloud/tools/jib/docker/DockerClient.java]: expanded

File: jib-core/src/main/java/com/google/cloud/tools/jib/docker/DockerClient.java (hunk 4 above)
Comment (gintautassulskus) [reply jib-core/src/main/java/com/google/cloud/tools/jib/docker/DockerClient.java]: addressed

File: jib-core/src/main/java/com/google/cloud/tools/jib/docker/DockerClient.java (hunk 5 above)
Comment (gintautassulskus) [reply jib-core/src/main/java/com/google/cloud/tools/jib/docker/DockerClient.java]: addressed

File: jib-core/src/test/java/com/google/cloud/tools/jib/docker/DockerClientTest.java (hunk 6 above)
Comment (gintautassulskus) [reply jib-core/src/test/java/com/google/cloud/tools/jib/docker/DockerClientTest.java]: addressed

File: jib-core/src/test/java/com/google/cloud/tools/jib/docker/DockerClientTest.java (hunk 7 above)
Comment (gintautassulskus) [reply jib-core/src/test/java/com/google/cloud/tools/jib/docker/DockerClientTest.java]: addressed

File: jib-core/src/test/java/com/google/cloud/tools/jib/docker/DockerClientTest.java (hunk 8 above)
Comment (gintautassulskus) [reply jib-core/src/test/java/com/google/cloud/tools/jib/docker/DockerClientTest.java]: addressed

File: jib-gradle-plugin/README.md (hunk 9 above)
Comment (gintautassulskus) [reply jib-gradle-plugin/README.md]: will do

File: jib-gradle-plugin/src/main/java/com/google/cloud/tools/jib/gradle/BuildDockerTask.java (hunk 10 above)
Comment (gintautassulskus) [reply jib-gradle-plugin/src/main/java/com/google/cloud/tools/jib/gradle/BuildDockerTask.java]: nope

File: jib-gradle-plugin/src/main/java/com/google/cloud/tools/jib/gradle/BuildDockerTask.java (hunk 11 above)
Comment (gintautassulskus) [reply jib-gradle-plugin/src/main/java/com/google/cloud/tools/jib/gradle/BuildDockerTask.java]: nope

File: jib-core/src/main/java/com/google/cloud/tools/jib/docker/DockerClient.java (hunk 13 above)
Comment (gintautassulskus) [reply jib-core/src/main/java/com/google/cloud/tools/jib/docker/DockerClient.java]: Yes, but I do not think that this is a good solution. 

We want to test the actual use case, which is creating the process builder using one of the defined `newClient(*)` methods. Then, however, the only way to access the private process builder object is using the `getProcessBuilderFactory` getter.

If we are going to create the object directly using `defaultProcessBuilderFactory()`, we will not test the whole execution path.

File: jib-core/src/test/java/com/google/cloud/tools/jib/docker/DockerClientTest.java (hunk 14 above)
Comment (gintautassulskus) [reply jib-core/src/test/java/com/google/cloud/tools/jib/docker/DockerClientTest.java]: addressed

File: jib-core/src/test/java/com/google/cloud/tools/jib/docker/DockerClientTest.java (hunk 15 above)
Comment (gintautassulskus) [reply jib-core/src/test/java/com/google/cloud/tools/jib/docker/DockerClientTest.java]: addressed

File: jib-core/src/test/java/com/google/cloud/tools/jib/docker/DockerClientTest.java (hunk 16 above)
Comment (gintautassulskus) [reply jib-core/src/test/java/com/google/cloud/tools/jib/docker/DockerClientTest.java]: addressed

File: jib-core/src/main/java/com/google/cloud/tools/jib/docker/DockerClient.java (hunk 13 above)
Comment (chanseokoh) [reply jib-core/src/main/java/com/google/cloud/tools/jib/docker/DockerClient.java]: What you said makes sense, but I think this is basically a matter of the scope of a unit test, like a small unit test vs. medium or large unit test (more toward an end-to-end integration test). It's OK and meaningful to have narrowly scoped unit tests that test only one method, and if it's possible, it's also meaningful to have another larger tests that cover a  longer execution path at a higher level. I just feel it doesn't seem worth adding a new method (especially a simple getter) just to enable a test. After all, what you get from the getter is just an object built once from `defaultProcessBuilderFactory` at construction.

Comment (chanseokoh) [reply jib-core/src/main/java/com/google/cloud/tools/jib/docker/DockerClient.java]: To add, I'm trying to avoid spreading the practice like, "oh, I cannot access these fields in tests, but I want to directly examine the values in tests. I'll just make them visible by adding getters." It feels like, you may be more focused on testing internal implementation or design that is meant to be encapsulated.

File: jib-maven-plugin/src/test/resources/projects/war_servlet25/src/main/java/example/HelloWorld.java [hunk 17]
```diff
@@ -16,16 +16,16 @@
 
 package example;
 
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
```
Comment (chanseokoh) [inline jib-maven-plugin/src/test/resources/projects/war_servlet25/src/main/java/example/HelloWorld.java]: nit: unnecessary changes in test data (and not in the Google style, actually).

File: jib-maven-plugin/src/test/resources/projects/simple/src/main/java/com/test/HelloWorld.java [hunk 18]
```diff
@@ -17,6 +17,7 @@
 package com.test;
 
 import dependency.Greeting;
+
```
Comment (chanseokoh) [inline jib-maven-plugin/src/test/resources/projects/simple/src/main/java/com/test/HelloWorld.java]: nit: unnecessary changes in test data (and not in the Google style, actually).

File: jib-maven-plugin/src/test/resources/projects/multi/lib/src/test/java/com/lib/LibTest.java [hunk 19]
```diff
@@ -16,8 +16,6 @@
 
 package com.lib;
 
-import static org.junit.Assert.assertTrue;
```
Comment (chanseokoh) [inline jib-maven-plugin/src/test/resources/projects/multi/lib/src/test/java/com/lib/LibTest.java]: nit: unnecessary changes in test data (and not in the Google style, actually).

File: jib-maven-plugin/src/test/resources/projects/default-target/src/main/java/com/test/HelloWorld.java [hunk 20]
```diff
@@ -17,6 +17,7 @@
 package com.test;
 
 import dependency.Greeting;
+
```
Comment (chanseokoh) [inline jib-maven-plugin/src/test/resources/projects/default-target/src/main/java/com/test/HelloWorld.java]: nit: unnecessary changes in test data (and not in the Google style, actually).

File: jib-core/src/main/java/com/google/cloud/tools/jib/api/DockerDaemonImage.java [hunk 21]
```diff
@@ -55,41 +53,30 @@ public static DockerDaemonImage named(String imageReference)
...
   }
 
-  /**
-   * Gets the path to the {@code docker} CLI.
-   *
-   * @return the path to the {@code docker} CLI
-   */
-  Path getDockerExecutable() {
-    return dockerExecutable;
+  public DockerClient getDockerClient() {
```
Comment (chanseokoh) [inline jib-core/src/main/java/com/google/cloud/tools/jib/api/DockerDaemonImage.java]: Looks like this exists only for testing. It'd be better to avoid this, but if you think you can't, at least add `@VisibleForTesting` for indication (even though it is already public).

File: jib-core/src/main/java/com/google/cloud/tools/jib/api/DockerDaemonImage.java [hunk 22]
```diff
@@ -55,41 +53,30 @@ public static DockerDaemonImage named(String imageReference)
...
+    return dockerClient;
+  }
+
+  public DockerDaemonImage setDockerClient(DockerClient dockerClient) {
```
Comment (chanseokoh) [inline jib-core/src/main/java/com/google/cloud/tools/jib/api/DockerDaemonImage.java]: This is a public-facing jib-core API, so let's add a Javadoc.

File: jib-core/src/main/java/com/google/cloud/tools/jib/docker/DockerClient.java (hunk 13 above)
Comment (gintautassulskus) [reply jib-core/src/main/java/com/google/cloud/tools/jib/docker/DockerClient.java]: I can use `defaultProcessBuilderFactory` but then the last assertion in `DockerDaemonImageTest` is no longer feasible.

File: jib-core/src/main/java/com/google/cloud/tools/jib/api/DockerDaemonImage.java (hunk 21 above)
Comment (gintautassulskus) [reply jib-core/src/main/java/com/google/cloud/tools/jib/api/DockerDaemonImage.java]: For now, I left it with @ VisibleForTesting

**Comment by gintautassulskus** [comment]
I replaced `setDockerExecutable` and `setDockerEnvironment` with `setDockerClient()` but now the user must set it like this:
`jibDockerBuild.dockerClient = com.google.cloud.tools.jib.docker.DockerClient.newClient()`

This needs to be changed to a closure. Until then, it's better not to merge this PR.

Unfortunately, I do not have time to work on this at the moment.

**Comment by chanseokoh** [comment]
> I replaced setDockerExecutable and setDockerEnvironment with setDockerClient() but now the user must set it like this:
> jibDockerBuild.dockerClient = com.google.cloud.tools.jib.docker.DockerClient.newClient()

I think we can retain `setDockerExecutable` and `setDockerEnvironment` in `DockerBuildTask`.

But anyway, now you said it, I just realized that `DockerClient` is not a public user-facing API (i.e., not under the `.api` package). Sorry for the turnaround. I should have been more careful.

And since you said you'll not have time for this, I think we can take over from now.

Thanks again for your great contribution. Much appreciated.

**Comment by gintautassulskus** [comment]
No prob. Thanks for your patience and attention.

**Comment by chanseokoh** [comment]
@GoogleContainerTools/java-tools I'll create a branch from this and put up a PR.

**Comment by chanseokoh** [comment]
@gintautassulskus we've merged #1214, which will make the Docker client parameters available in the next 0.10.0 release. Thanks a ton for your contribution!

**Comment by gintautassulskus** [comment]
Thanks for taking over and finishing it.
